Initial Engine implementation and DB changes

HarveyHunt opened the pull request:
This is totally untested (I don't think the sort function will even work), but I'll write some tests for it later.

There is some more DB stuff to do as well as WS, I think manfred and I will send a PR together...

HarveyHunt committed:
engine: Initial implementation

Signed-off-by: Harvey Hunt <harveyhuntnexus@gmail.com>

manfredlift committed:
db: Query for available parking lots within distance

HarveyHunt committed:
db: Add lot and allocation getters

HarveyHunt committed:
db: Add function to delete allocations

Signed-off-by: Harvey Hunt <harveyhuntnexus@gmail.com>

manfredlift commented on parking/backend/engine/alloc_engine.py:
```diff
@@ -0,0 +1,54 @@
+        return success
+
+    async def remove_allocation(self, user_id: str) -> None:
+        self.user_sessions.deallocate(user_id)
```
Missing 'await'.

> Calling a coroutine does not start its code running – the coroutine object returned by the call doesn’t do anything until you schedule its execution.

HarveyHunt replied:
Fixed :+1:

manfredlift commented on parking/backend/engine/alloc_engine.py:
```diff
@@ -0,0 +1,54 @@
+            logger.warning('Failed to allocate parking lot {} to user {}'.format(park_id, user_id))
+        return success
+
+    async def remove_allocation(self, user_id: str) -> None:
```
I also need a 'delete_allocation(user_id)' function to call from the Websocket handler.

HarveyHunt replied:
Done :+1:

manfredlift commented on parking/backend/engine/alloc_engine.py:
```diff
@@ -0,0 +1,54 @@
+
+    async def handle_allocation_request(self, user_id: str, request: ParkingRequestMessage) -> Optional[ParkingLot]:
+        rejections = self.user_sessions.get_user(user_id).rejections
+        lots = await self.dba.get_available_parking_lots(request.location, 100, rejections)
```
Try to get 'distance' from ParkingRequestMessage.prefereces['distance'], if not then use a default value larger than 100 meters.

HarveyHunt replied:
I'll check for this, it might be nice to create a model for preferences in the future.

I think that ```handle_allocation_message``` should probably raise an ```InvalidDistanceException``` that the caller will handle - thoughts?

bensimner replied:
Also isn't this call inconsistent with the signature given earlier:

    async def get_available_parking_lots(self, lat: float, long: float,
                                         dist_meters: int, exclusions: List[int]) -> List[Record]:

HarveyHunt replied:
Yeah, that's fixed in the newest version - we went with passing a ```Location``` object instead.

bensimner replied:
I see, is the existence in this commit an artefact of your rebasing shenanigans by any chance

HarveyHunt replied:
Surprisingly not ;) It's due to us discussing the interface and then forgetting it, whilst having no tests to verify if it worked. I don't know which is worse really...

FlinnBurgess commented on parking/backend/db/dbaccess.py:
```diff
@@ -1,8 +1,9 @@
 import asyncpg
+from asyncpg import Record
 
 import parking.backend.db.sql_constants as c
```
Your use of single-letter variables is offensive to my eyes and my brain

manfredlift replied:
(Written by me). This has to with preventing massive line lengths. sql_constants.py constants have really long names already and sql_constants.py is only used internally by dbaccess.py so it shouldn't be that offensive.

FlinnBurgess replied:
I can understand the reasoning behind having the sql constants and my comment doesn't really matter massively for this project, but you should write code in a way that benefits anyone who comes after you. `c` might be nice and easy for you to type but it's essentially meaningless, something like `sql_strings` would be a lot more readable. Again, not that I'm expecting you to change it!

Chippiewill replied:
Variable naming is always a delicate balance. I think in this instance it's ok as while `c` itself does not convey significant meaning by itself, it does in conjunction with the variables in that namespace (e.g. the purpose of `c.PARKINGLOTS_SELECT` is obvious - especially in the context it's used in)

manfredlift commented on parking/backend/engine/alloc_engine.py:
```diff
@@ -0,0 +1,54 @@
+        self.dba = dba
+        self.user_sessions = user_sessions
+
+    async def handle_allocation_request(self, user_id: str, request: ParkingRequestMessage) -> Optional[ParkingLot]:
```
Please return 'ParkingLotAllocation' object instead I created in ws_models. It includes the distance that should be displayed to the user:
```
@attr.s
class ParkingLotAllocation(ParkingLot):
    distance: int = attr.ib(validator=enforce_type, default=0)
```

FlinnBurgess commented on parking/backend/db/dbaccess.py:
```diff
@@ -44,6 +46,11 @@ class DbAccess(object):
+            parking_lot: Record = await conn.fetchrow(c.PARKINGLOTS_SELECT, park_id)
+        return parking_lot
+
     async def delete_parking_lot(self, park_id: int) -> Optional[int]:
```
Can't say I like the return values of these functions. You'd expect a success or failure if you were trying to delete something, but instead you're getting either some random ID or a null value? Not only are the return types different but they also seem pretty irrelevant.

manfredlift replied:
This has to do with consistency. Since insert_parking_lot needs to return an id or nothing, we chose all other updates/deletes to return the same thing. I agree that the alternative would be true/false.

FlinnBurgess replied:
Then my argument would be that this is consistently bad practice!! Wouldn't it be better to return a tuple or something with the status and the relevant data? e.g. (success = True, data = new_id) or (success = false, data = null)

bensimner replied:
Usually, in python, invalid states are flagged by raising exceptions.

If we assume nobody will try delete a parking lot that doesn't exist, then this is clearly an error and should raise an exception.

Either way, it's better to be consistent about these things than for some to be one way in one place and others to be a different way in other places.

Chippiewill replied:
This is again a delicate balance. For something like `delete` there are a few possibilities:

- The delete succeeds: The callee requires no further information (other than the coroutine returns)
- The delete fails because it didn't exist: The callee needs to be informed that it didn't exist
- The delete fails because of a general DB error (e.g. query took too long): The callee needs to be informed of DB errors

For the first possibility the function can just return nothing (the id is meaningless, especially as the entry no longer exists). For the third possibility we can consider this as a firmly "exceptional" circumstance so an exception should definitely be raised (which I imagine asyncpg will handle for us).

The second possibility is more nuanced. The id of the lot being deleted does not come from within the program but from external input therefore we cannot say it is "exceptional" or an error for this to occur, for instance two deletes for the same lot may come in at the same time, so one might be inclined to use a boolean return value to indicate failure. That being said the "exceptional" criteria does not apply to Python. Exceptions _are_ recommended for control flow in pythonic code (within reason) and I think in this instance would be suitable as we ultimately need to propagate this up as a 404 response.

This is basically a longwinded way of saying I agree with Ben.

FlinnBurgess commented on parking/backend/db/dbaccess.py:
```diff
@@ -74,3 +81,26 @@ class DbAccess(object):
+            records: List[Record] = await conn.fetch(c.ALLOCATIONS_SELECT, park_id)
+        return records
+
+    async def delete_allocation(self, user_id: str) -> None:
```
Again, no indication of whether this actually works or not. How are you validating this stuff?

manfredlift replied:
There are actual tests in tests/test_db.py?

FlinnBurgess replied:
And what if something goes wrong that is outside the scope of the unit test?

FlinnBurgess commented on parking/backend/db/sql_constants.py:
```diff
@@ -63,6 +80,50 @@
+        ParkingLots
+    WHERE
+        earth_box(ll_to_earth($1, $2), $3) @> ll_to_earth(lat, long) 
+        AND num_available > num_allocated
```
num_available is a bit of a misleading name, since num_allocated of them are in fact not available. Should be something like capacity, with the order of the equality reversed:
`num_allocated < capacity`
This is a lot more readable in my opinion

manfredlift replied:
In this system:
capacity = total parking spaces in a lot
num_available = number of parking spaces available as per sensors
num_allocated = number of parking spaces allocated by our system

We need all of these to make decisions and `num_allocated < capacity` wouldn't work.

FlinnBurgess replied:
Okay, that makes more sense

FlinnBurgess commented on parking/backend/engine/alloc_engine.py:
```diff
@@ -0,0 +1,54 @@
+        self.user_sessions = user_sessions
+
+    async def handle_allocation_request(self, user_id: str, request: ParkingRequestMessage) -> Optional[ParkingLot]:
+        rejections = self.user_sessions.get_user(user_id).rejections
```
rejections is a liiiiiittle bit vague. I'm guessing it's parking lots that the user has rejected, rather than rejections of the user by the engine?

manfredlift replied:
Yes, could add a comment or rename to user_rejections.

HarveyHunt replied:
I'll rename to user_rejections

FlinnBurgess commented on parking/backend/engine/alloc_engine.py:
```diff
@@ -0,0 +1,54 @@
+            lot = lots[0]
+            return ParkingLot(lot['capacity'], lot['name'], lot['price'], Location(lot['lat'], lot['long']))
+        else:
+            return lots[0] if lots else None
```
I'm confused by this line, surely it would never get executed if lots was non-null, so what's with the check?

HarveyHunt replied:
That's an editing error, should just be ```return None```

FlinnBurgess commented on parking/backend/engine/alloc_engine.py:
```diff
@@ -0,0 +1,54 @@
+        # Just take the first lot for now...
+        if lots:
+            lot = lots[0]
+            return ParkingLot(lot['capacity'], lot['name'], lot['price'], Location(lot['lat'], lot['long']))
```
It would be a lot nicer if you could just pass `lot` into the constructor and get an object, rather than passing each array element in... A quick google brings up classmethods, what are your thoughts on this? 
https://stackoverflow.com/questions/141545/how-to-overload-init-method-based-on-argument-type

HarveyHunt replied:
It'd be nice to have a better way of constructing the object, a classmethod could be a good idea. I'll think about it more tomorrow and will have something better for the final PR

FlinnBurgess commented on parking/backend/engine/alloc_engine.py:
```diff
@@ -0,0 +1,54 @@
+    async def recalculate_allocations(self, park_id: int) -> None:
+        # Deallocate users if parking spaces are taken
+        # Calculate the allocated users and find the furthest away
+        lot = await self.dba.get_parking_lot(park_id)
```
`lot_id` seems more appropriate than `park_id`

HarveyHunt replied:
Whilst I agree, the DB stuff is using ```park_id``` so I was remaining consistent with that

FlinnBurgess commented on parking/backend/engine/alloc_engine.py:
```diff
@@ -0,0 +1,54 @@
+        # Calculate the allocated users and find the furthest away
+        lot = await self.dba.get_parking_lot(park_id)
+        user_sessions = self.user_sessions
+        overflow = lot.num_allocated - lot.num_available
```
Just to reiterate on the `num_available` column name. Also, it would probably be more readable to use a variable name like `over_allocation_occured` and do the inequality on this line, then later you can do `if over_allocation_occured`

HarveyHunt replied:
The ```overflow``` value gets used later though, so converting it to a bool immediately wouldn't help much. I think that ```overflow > 0``` is quite readable tbh

FlinnBurgess commented on tests/test_db.py:
```diff
@@ -138,3 +139,83 @@ def teardown_module(module):
+
+
+@pytest.mark.asyncio
+async def test_get_available_parking_lots(event_loop):
```
I'm guessing these tests work because of your choice of location coordinates, you might want to give a brief explanation of this? It's not immediately obvious to me who doesn't know the code all that well.

manfredlift replied:
This is due to 1 degree of latitude equaling to about 110km and the distance of 1 degree of longitude depends on the exact latitude. Some sanity checks and comparisons were done with some online calculators as well to choose a test case.

manfredlift replied:
@uptonfunk, this reminds me that the simulation team should be aware that we are using earth coordinates at the backend to calculate distances.

FlinnBurgess commented on tests/test_db.py:
```diff
@@ -138,3 +139,83 @@ def teardown_module(module):
+
+        park_id = await db.insert_parking_lot(parking_lot)
+        result = await db.allocate_parking_lot(user_id, park_id)
+        assert result
```
Some of your assertions seem to be irrelevant to the name of your test. To me this suggests they should be in their own tests. You've already tested this above, so it seems pointless to put it here as well.

HarveyHunt replied:
True, I used lots of asserts to make debugging easier if the test fails in the future.

FlinnBurgess reviewed (commented):
Perhaps a little nit-picky at times, let me know if this is helpful feedback

Chippiewill commented on parking/backend/db/dbaccess.py:
```diff
@@ -74,3 +81,26 @@ class DbAccess(object):
+            async with conn.transaction():
+                park_id: int = await conn.fetchval(c.ALLOCATIONS_DELETE, user_id)
+                if park_id:
+                    await conn.fetchval(c.PARKINGLOTS_DECREMENT_ALLOCATION, park_id)
```
Big issue here as you're doing your delete and decrement in separate database transactions which can result in an inconsistent database. You should set up an explicit transaction that does both queries at once using `BEGIN`/`COMMIT`.

HarveyHunt replied:
The context manager (```async with conn.transaction()```) creates the transaction that the operations are performed in.

I realised that I can use ```await conn.execute(c.PARKINGLOTS_DECREMENT_ALLOCATION, park_id)``` instead of fetchval so have switched it

Chippiewill commented on parking/backend/db/dbaccess.py:
```diff
@@ -74,3 +81,26 @@ class DbAccess(object):
+                                         dist_meters: int, exclusions: List[int]) -> List[Record]:
+        async with self.pool.acquire() as conn:
+            if exclusions:
+                print(exclusions)
```
Debug

HarveyHunt committed:
engine: PR fixes

Signed-off-by: Harvey Hunt <harveyhuntnexus@gmail.com>

HarveyHunt committed:
engine: Add delete_allocation

Used to remove an allocation from the DB but without disconnecting the
client in UserSessions.

Signed-off-by: Harvey Hunt <harveyhuntnexus@gmail.com>

HarveyHunt committed:
db: Convert fetchval -> execute

Signed-off-by: Harvey Hunt <harveyhuntnexus@gmail.com>

bensimner commented on parking/backend/engine/alloc_engine.py:
```diff
@@ -0,0 +1,58 @@
+
+
+class AllocationEngine():
+    def __init__(self, dba: DbAccess, user_sessions: 'UserSessions') -> None:
```
Why is `UserSessions` in quotations here rather than just imported?

HarveyHunt replied:
There was a circular dependency between engine and wsserver (where UserSessions is).

@manfredlift has now split UserSessions into its own file, so this won't be an issue now.

bensimner commented on parking/backend/engine/alloc_engine.py:
```diff
@@ -0,0 +1,58 @@
+logger = logging.getLogger('backend')
+
+
+class AllocationEngine():
```
Should the engine handle all connections to the DbAccess? 

i.e. anything going through the REST API should go through the `AllocationEngine` ?

HarveyHunt replied:
The engine is only going to be used for allocations, so I think it's fine for the REST API to communicate with the DB directly.
